pkg/server/plugin/noderesolver/aws/iid.go

```diff
@@ -0,0 +1,282 @@
+	return &spi.GetPluginInfoResponse{}, nil
+}
+
+func (p *IIDResolverPlugin) Resolve(ctx context.Context, req *noderesolver.ResolveRequest) (*noderesolver.ResolveResponse, error) {
+	resp := &noderesolver.ResolveResponse{
+		Map: make(map[string]*common.Selectors),
+	}
+	for _, spiffeID := range req.BaseSpiffeIdList {
+		selectors, err := p.resolveSpiffeID(ctx, spiffeID)
+		if err != nil {
+			return nil, err
+		}
+		if selectors != nil {
+			resp.Map[spiffeID] = selectors
+		}
+	}
+	return resp, nil
+}
+
+func (p *IIDResolverPlugin) resolveSpiffeID(ctx context.Context, spiffeID string) (*common.Selectors, error) {
+	_, region, instanceID, err := parseAgentID(spiffeID)
+	if err != nil {
+		logrus.Warnf("unrecognized Agent ID: %s: %v", spiffeID, err)
+		return nil, nil
+	}
+
+	client, err := p.getRegionClient(region)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := client.DescribeInstancesWithContext(ctx, &ec2.DescribeInstancesInput{
+		InstanceIds: []*string{aws.String(instanceID)},
+		Filters:     instanceFilters,
+	})
+	if err != nil {
+		return nil, iidError.Wrap(err)
+	}
+
+	selectorSet := map[string]bool{}
+	addSelectors := func(values []string) {
+		for _, value := range values {
+			selectorSet[value] = true
+		}
+	}
+
+	for _, reservation := range resp.Reservations {
+		for _, instance := range reservation.Instances {
+			addSelectors(resolveTags(instance.Tags))
+			addSelectors(resolveSecurityGroups(instance.SecurityGroups))
+			if instance.IamInstanceProfile != nil && instance.IamInstanceProfile.Arn != nil {
+				output, err := client.GetInstanceProfileWithContext(ctx, &iam.GetInstanceProfileInput{
+					InstanceProfileName: instance.IamInstanceProfile.Arn,
+				})
+				if err != nil {
+					return nil, iidError.Wrap(err)
+				}
+				addSelectors(resolveInstanceProfile(output.InstanceProfile))
+			}
+		}
+	}
+
+	// sort and dedup selectors
+	values := make([]string, 0, len(selectorSet))
+	for value := range selectorSet {
+		values = append(values, value)
+	}
+	sort.Strings(values)
+
+	selectors := new(common.Selectors)
+	for _, value := range values {
+		selectors.Entries = append(selectors.Entries, &common.Selector{
+			Type:  pluginName,
+			Value: value,
+		})
+	}
+
+	return selectors, nil
+}
+
+func (p *IIDResolverPlugin) getRegionClient(region string) (awsClient, error) {
+	// do an initial check to see if p client for this region already exists
+	p.mu.RLock()
+	client, ok := p.clients[region]
+	p.mu.RUnlock()
+	if ok {
+		return client, nil
+	}
+
+	// no client for this region. make one.
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	// more than one thread could be racing to create p client (since we had
+	// to drop the read lock to take the write lock), so double check somebody
+	// hasn't beat us to it.
+	client, ok = p.clients[region]
+	if ok {
+		return client, nil
+	}
+
+	if p.conf == nil {
+		return nil, iidError.New("not configured")
+	}
+	p.conf.Region = aws.String(region)
```

Comment:
conf is a pointer to aws config, and we mutate that config for every client we generate... is it safe to mutate the region out from under previously-generated clients? also feels a little icky to have the conf be set on the struct and the value of Region equal to whichever client was last fetched... maybe instead we can keep `IIDResolverConfig` on the struct, and form `aws.Config` structs as needed in `newAWSClient`

Reply:
sure, i'm happy to do that. for what it's worth, the config values are copied when the new session is created. I agree it looks surprising, so totally worth fixing to me.